src/simulators/matrix_product_state/matrix_product_state_internal.cpp

```diff
@@ -1586,6 +1543,115 @@ void MPS::propagate_to_neighbors_internal(uint_t min_qubit, uint_t max_qubit,
+  double prob = 1;
+  std::string current_measure="";
+  char measure_1_qubit;
+  bool is_first_qubit = true;
+  cmatrix_t mat;
+
+  for (uint_t i=0; i<size; i++) {
+    measure_1_qubit = sample_measure_single_qubit(i, is_first_qubit, current_measure, 
+						  prob, rnds[i], mat);
+    current_measure =  measure_1_qubit + current_measure;
+    is_first_qubit = false;
+  }
+  return  create_outcome_vector(qubits, current_measure);
+}
+
+uint_t MPS::sample_measure_single_qubit(uint_t qubit,
+					bool is_first_qubit,
+					std::string &prev_measure, 
+					double &prob, double rnd,
+					cmatrix_t &mat) const {
+  double prob0 = 0;
+  if (is_first_qubit) {
+    reg_t qubits_to_update;
+    qubits_to_update.push_back(qubit);
+    // step 1 - measure qubit in Z basis
+    double exp_val = real(expectation_value_pauli_internal(qubits_to_update, "Z", qubit, qubit, 0));
+    // step 2 - compute probability for 0 or 1 result
+    prob0 = (1 + exp_val ) / 2;
+  } else {
+    std::string new_string = '0' + prev_measure;
+    prob0 = get_single_probability0(qubit, mat);
+    prob0 /= prob;
+  }
+
+  char measurement = (rnd < prob0) ? '0' : '1';
+  double new_prob = (measurement == '0') ? prob0 : 1-prob0;
+  prob *= new_prob;
+
+  // Now update mat for the next qubit
+  // mat represents the accumulated product of the matrices of the current
+  // measurement outcome
+  uint_t bit = measurement == '0'? 0 : 1;
+  if (is_first_qubit) {
+    mat = q_reg_[qubit].get_data(bit);
+    if (qubit != 0)  // multiply mat by left lambda
+      for (uint_t col=0; col<mat.GetColumns(); col++)
+	for (uint_t row=0; row<mat.GetRows(); row++)
+	  mat(row, col) *= lambda_reg_[qubit-1][row];
+  } else {
+    mat = mat * q_reg_[qubit].get_data(bit);
+  }
+  if (qubit != num_qubits_-1) {  // multiply mat by right lambda
+    for (uint_t row=0; row<mat.GetRows(); row++)
+      for (uint_t col=0; col<mat.GetColumns(); col++)
+	mat(row, col) *= lambda_reg_[qubit][col];
+  }
+  return measurement;
+}
+
+double MPS::get_single_probability0(uint_t qubit, const cmatrix_t &mat) const {
+  // multiply by the matrix for measurement of 0
+  cmatrix_t temp_mat = mat * q_reg_[qubit].get_data(0);
+
+  if (qubit != num_qubits_-1) {
+    for (uint_t row=0; row<temp_mat.GetRows(); row++) {
+      for (uint_t col=0; col<temp_mat.GetColumns(); col++) {
+	temp_mat(row, col) *= lambda_reg_[qubit][col];
+      }
+    }
+  }
+  //prob0 = the probability to measure 0
+  double prob0 = real(AER::Utils::sum( AER::Utils::elementwise_multiplication(temp_mat, AER::Utils::conjugate(temp_mat))));
+  return prob0;
+}
+
+reg_t MPS::create_outcome_vector(const reg_t &qubits, const std::string &current_measure) const {
+  uint_t size = qubits.size();
+  reg_t outcome_vector(size), ordered_outcome(size), final_outcome(size);
+
+  // copy string of outcome to vector while reversing the direction of the qubits
+  for (uint_t i=0; i<size; i++) {
+    outcome_vector[size-1-i] = (current_measure[i] == '0') ? 0 : 1;
```

Comment:
What's the benefit of working with a string all the time, and not directly on the vector?

Reply:
You are right - good comment! I was simply caught up in thinking of measurement as a string. Changed the string to a reg_t. This simplifies the code.